src/cose/sign1.rs

```diff
@@ -0,0 +1,484 @@
+    #[error("error occurred when signing COSE_Sign1: {0}")]
+    Signing(signature::Error),
+    #[error("unable to set ClaimsSet: {0}")]
+    UnableToDeserializeIntoClaimsSet(CoseError),
+}
+
+/// Result with error type: [`Error`].
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Result for verification of a COSE_Sign1.
+#[derive(Debug)]
+pub enum VerificationResult {
+    Success,
+    Failure(String),
+    Error(Error),
+}
+
+impl VerificationResult {
+    /// Result of verification.
+    ///
+    /// `false` implies the signature is inauthentic or the verification algorithm encountered an
+    /// error.
+    pub fn is_success(&self) -> bool {
+        matches!(self, VerificationResult::Success)
+    }
+
+    /// Translate to a std::result::Result.
+    ///
+    /// Converts failure reasons and errors into a String.
+    pub fn into_result(self) -> Result<(), String> {
+        match self {
+            VerificationResult::Success => Ok(()),
+            VerificationResult::Failure(reason) => Err(reason),
+            VerificationResult::Error(e) => Err(format!("{}", e)),
+        }
+    }
+
+    /// Retrieve the error if the verification algorithm encountered an error.
+    pub fn into_error(self) -> Option<Error> {
+        match self {
+            VerificationResult::Error(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl PreparedCoseSign1 {
+    pub fn new(
+        builder: coset::CoseSign1Builder,
+        detached_payload: Option<&[u8]>,
+        aad: Option<&[u8]>,
+        tagged: bool,
+    ) -> Result<Self> {
+        let cose_sign1 = builder.build();
+
+        // Check if the payload is present and if it is attached or detached.
+        // Needs to be exclusively attached or detached.
+        let payload = match (cose_sign1.payload.as_ref(), detached_payload) {
+            (Some(_), Some(_)) => return Err(Error::DoublePayload),
+            (None, None) => return Err(Error::NoPayload),
+            (Some(payload), None) => payload.clone(),
+            (None, Some(payload)) => payload.to_vec(),
+        };
+        // Create the signature payload ot be used later on signing.
+        let signature_payload = sig_structure_data(
+            SignatureContext::CoseSign1,
+            cose_sign1.protected.clone(),
+            None,
+            aad.unwrap_or_default(),
+            &payload,
+        );
+
+        Ok(Self {
+            cose_sign1: MaybeTagged::new(tagged, cose_sign1),
+            signature_payload,
+        })
+    }
+
+    /// Returns the signature payload that needs to be used to sign.
+    pub fn signature_payload(&self) -> &[u8] {
+        &self.signature_payload
+    }
+
+    /// Finalize the COSE_Sign1 by adding the signature.
+    pub fn finalize(self, signature: Vec<u8>) -> MaybeTagged<CoseSign1> {
+        let mut cose_sign1 = self.cose_sign1;
+        cose_sign1.inner.signature = signature;
+        cose_sign1
+    }
+}
+
+impl MaybeTagged<CoseSign1> {
```

Comment:
Same comment as for CoseMac0
```suggestion
impl CoseSign1 {
```

Reply:
same as above cannot impl for outside struct